File: bindings/python/examples/how_tos/native_tokens/melt.py
import os

from dotenv import load_dotenv
from iota_sdk import Wallet, WalletOptions

load_dotenv()

# In this example we will decrease the native token supply

wallet = Wallet(WalletOptions(storage_path=os.environ.get('WALLET_DB_PATH')))

# Sync wallet with the node
balance = wallet.sync()

# Find first foundry and corresponding token id
token_id = balance.foundries[0]

available_balance = balance.native_tokens[token_id].available
print(f'Balance before melting: {available_balance}')

if 'STRONGHOLD_PASSWORD' not in os.environ:
    raise Exception(".env STRONGHOLD_PASSWORD is undefined, see .env.example")

wallet.set_stronghold_password(os.environ["STRONGHOLD_PASSWORD"])

melt_amount = 10

# Send transaction.
transaction = wallet.melt_native_token(token_id, melt_amount)
print(f'Transaction sent: {transaction.transaction_id}')

# Wait for transaction to get accepted
block_id = wallet.wait_for_transaction_acceptance(
    transaction.transaction_id)
print(f'Tx accepted in block: {os.environ["EXPLORER_URL"]}/block/{block_id}')

balance = wallet.sync()
available_balance = balance.native_tokens[token_id].available
print(f'Balance after melting: {available_balance}')

File: bindings/nodejs/examples/how_tos/native_tokens/melt.ts
// Copyright 2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { getUnlockedWallet } from '../../wallet/common';

// The amount of native tokens to melt.
const MELT_AMOUNT = BigInt(10);

// In this example we will melt an existing native token with its foundry.
//
// Make sure that `STRONGHOLD_SNAPSHOT_PATH` and `WALLET_DB_PATH` already exist by
// running the `how_tos/wallet/create-wallet` example!
//
// Rename `.env.example` to `.env` first, then run
// yarn run-example ./how_tos/native_tokens/melt.ts
async function run() {
    for (const envVar of ['EXPLORER_URL']) {
        if (!(envVar in process.env)) {
            throw new Error(`.env ${envVar} is undefined, see .env.example`);
        }
    }
    try {
        // Create the wallet
        const wallet = await getUnlockedWallet();

        // May want to ensure the wallet is synced before sending a transaction.
        let balance = await wallet.sync();

        if (balance.foundries.length == 0) {
            throw new Error(`No Foundry available in account 'Alice'`);
        }

        // Find first foundry and corresponding token id
        const tokenId = balance.foundries[0];

        let token = balance.nativeTokens[tokenId];
        if (token == null) {
            throw new Error(
                `Couldn't find native token '${tokenId}' in the wallet`,
            );
        }

        console.log(`Balance before melting: ${token.available}`);

        // Melt some of the circulating supply
        const transaction = await wallet.meltNativeToken(tokenId, MELT_AMOUNT);

        console.log(`Transaction sent: ${transaction.transactionId}`);

        // Wait for transaction to get accepted
        await wallet.waitForTransactionAcceptance(transaction.transactionId);

        console.log(
            `Tx accepted: ${process.env.EXPLORER_URL}/transactions/${transaction.transactionId}`,
        );

        balance = await wallet.sync();
        token = balance.nativeTokens[tokenId];
        if (token == null) {
            throw new Error(
                `Couldn't find native token '${tokenId}' in the wallet`,
            );
        }
        console.log(`Balance after melting: ${token.available}`);
    } catch (error) {
        console.log('Error: ', error);
    }
    process.exit(0);
}

void run().then(() => process.exit());
